[Step2] 문자열 계산기

namonak opened the pull request:
이동철(@sah3122) 리뷰어님,

저의 첫? PR 요청이 좀 늦었습니다.. ㅎㅎ
혹독한 리뷰 부탁드립니다~ 감사합니다 :)

namonak committed:
refactor: add the step1 package and move the Person class to main

namonak committed:
test: add test code using Kotest

namonak committed:
feat: add test code and production code for the arithmetic function

namonak committed:
feat: add calculation functions

The following have been implement and test:
  - implement a method that verifies whether the input operand value is an integer
  - implement a method that returns a calculation result with input operands and operators
  - if the input operator is incorrect, raise an IllegalArgumentException

sah3122 commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,65 @@
+    println(StringCalculator.calculate(input))
+}
+
+class StringCalculator(input: String) {
```
StringCalculator 가 너무 많은 일을 하고 있는것 같아요.
문자열 파싱 / 계산과 같은 비즈니스에 대한 책임을 나누어보는건 어떨까요 ?

namonak replied:
코멘트 참고하여 아래와 같이 클래스를 나눠보았습니다.
- domain/Operand : 피연산자의 값을 저장
- domain/Calculator : 계산 기능을 처리
- StringCalculator : 입력갑 검증 및 문자열 파싱

sah3122 commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        private fun startCalculation(operator: String?, result: Int, operand: Int): Int = when (operator) {
+            "+" -> add(result, operand)
+            "-" -> subtract(result, operand)
+            "*" -> multiply(result, operand)
+            "/" -> divide(result, operand)
+            else -> throw IllegalArgumentException("연산자가 올바르지 않습니다.")
```
Enum을 활용하여 계산에 대한 책임을 전가해보세요 😄

namonak replied:
조언해 주신 내용 참고하여, 계산에 대한 기능을 enum class로 구현해 보았습니다!
확인 부탁드립니다 😊 

- https://github.com/next-step/kotlin-racingcar/commit/75c000c0be9203dd615f0d3026a81d8293a4c47f

sah3122 commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,65 @@
+            else -> throw IllegalArgumentException("연산자가 올바르지 않습니다.")
+        }
+
+        fun validateOperand(operand: String) : Int {
```
음수에 대한 유효성 검사도 필요해보입니다. 
양수를 보장하는 객체를 만들어보는것도 좋은 방법이 될것 같아요.

namonak replied:
조언해 주신 내용을 적용해 보기 위해, 아래와 같이 피연산자 값을 저장하는 Operand 클래스를 추가해 보려고 하는데요! 😊 
한가지 고민이 되는 것이 연산을 하기 위해서는, 결국 Operand 클래스에 getter 를 추가해야 할것 같은데요
혹시 getter를 추가하지 않고 피연자 값을 조회할 수 있는 방법은 없는 것인지 궁금합니다 ㅠㅠ

```kotlin
package step2.domain

class Operand(operand: Int) {
    init {
        require(operand >= 0) { "피연산자는 0 이상이어야 합니다." }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        return true
    }

    override fun hashCode(): Int {
        return javaClass.hashCode()
    }
}
```

namonak replied:
일단 남겨주신 코멘트 참조하여 아래와 같이 Operand 클래스와 OperandTest 테스트 코드를 추가해 보았습니다. 😊 

- https://github.com/next-step/kotlin-racingcar/commit/6ceaf7496c31fa320023e4ac3fdc108054782b6f

sah3122 commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,65 @@
+            var result = validateOperand(st.nextToken())
+
+            while (st.hasMoreTokens()) {
+                val operator = st.nextToken()
```
반복문 내부에서 커서를 변경하는것은 어떤 위험이 있을까요 ?
개인적으로는 연산자 / 피연산자에 대한 규칙을 활용하여 구현할것 같습니다 😄

namonak replied:
제 기존 구현대로 하는 경우, 두번째 `st.nextToken()` 호출때 `NoSuchElementException` 이 발생할 수 있는 점을 간과하였습니다.
좋은 지적 감사합니다. 😊 수정해 보도록 하겠습니다!

namonak replied:
좋은 방법은 아닌것 같지만, 반복분을 돌리기 원소의 갯수를 체크하는 기능을 추가해 보았습니다. 😭 

- https://github.com/next-step/kotlin-racingcar/commit/667a3a25342b1b84884b0e34483c727d7b601cb7

sah3122 commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,65 @@
+        }
+
+        fun calculate(s: String) : Int {
+            val st = StringTokenizer(s, " ")
```
매직넘버는 의미를 가지는 상수로 분리하여 관리해보세요 ! 
https://www.slipp.net/questions/356

namonak replied:
전달해 주신 링크 참조하여 수정해 보았습니다! 확인 부탁드립니다! 😊 

- https://github.com/next-step/kotlin-racingcar/commit/b672de40b3614839019f8c69c5aaefba84961d52

sah3122 commented on src/test/kotlin/step2/StringCalculatorTest.kt:
```diff
@@ -0,0 +1,83 @@
+import org.junit.jupiter.params.provider.CsvSource
+import org.junit.jupiter.params.provider.ValueSource
+
+class StringCalculatorTest {
```
상세한 테스트 코드 정의 👍

namonak replied:
칭찬 감사합니다! 😊

생각해 보니 첫 강의때 Kotest도 소개해 주셨는데, 이번에는 JUnit만 사용해서 테스트 코드를 작성했네요...
추후에는 Kotest를 이용해서 테스트 코드 작성하는 연습도 해보겠습니다!

sah3122 replied:
다양한 시도 좋습니다 !

sah3122 reviewed (changes requested):
문자열계산기 구현 잘해주셨네요 👍 
몇가지 고민거리를 남겨두었는데 확인 부탁드려요 🙏

namonak committed:
style: apply the lint report

namonak committed:
refactor: manage the input delimiter magic number as a constant

namonak committed:
feat: add an Operand class for operand values

The following have been implement and test:
 - throw an IllegalArgumentException when a negative number is input
 - create equals() and hashCode() methods

namonak committed:
feat: add a Calculator enum class for the calculation function

namonak committed:
feat: add exception handling for the number of input elements

sah3122 commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,43 @@
+package step2
+
+import step2.domain.Calculator
+import java.util.StringTokenizer
```
koltin 스러운 코드를 작성하기 위해 가장 접근하기 쉬운 방법으로는 java 관련 import 문을 없애보는것입니다. 
StringTokenizer 를 사용하기 보다는 입력 받은 문자열을 파싱하여 연산자 / 피연산자를 전달할수 있는 클래스를 정의해보는것도 좋을것 같아요 😄

namonak replied:
넵, 조언해 주신대로 `StringTokenizer`는 제거하였습니다. java 관련 패키지를 import 하지 않도록 주의하겠습니다! 😄 

- https://github.com/next-step/kotlin-racingcar/commit/ecc6214932b5660ef953d6df11e33974b49b5d12

sah3122 commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,43 @@
+
+    companion object {
+        fun calculate(s: String): Int {
+            val st = StringTokenizer(s, INPUT_DELIMITER)
```
```suggestion
            val tokens = StringTokenizer(s, INPUT_DELIMITER)
```
이번 과정을 진행하면서는 약어를 사용하기 보다는 변수명을 풀어서 선언해보세요. 
가독성이 좋아지는 장점이 있습니다 😄

namonak replied:
네, 조언해 주신대로 변수명을 풀어서 선언하겠습니다. `StringTokenizer`를 사용하지 않도록 수정하였고, `st` 변수는 삭제되었습니다. 😃 

- https://github.com/next-step/kotlin-racingcar/commit/ecc6214932b5660ef953d6df11e33974b49b5d12

sah3122 commented on src/main/kotlin/step2/StringCalculator.kt:
```diff
@@ -0,0 +1,43 @@
+            return result
+        }
+
+        fun validateOperand(operand: String): Int {
+            try {
+                return operand.toInt()
+            } catch (ex: NumberFormatException) {
+                throw NumberFormatException("숫자가 아닙니다.")
+            }
+        }
+    }
```
validate 라는 메소드명이 기능과 적절하게 매칭되지 않네요 ㅎㅎ
입력받은 문자열을 숫자로 변환하는 메소드명을 고려해보세요. 

추가로 클래스 내부에서만 사용하는 메소드는 private 메소드로 선언하는것도 좋아보입니다 :)

namonak replied:
https://github.com/next-step/kotlin-racingcar/pull/775#discussion_r1017244109 에서 조언해 주신대로, Operand 클래스의 보조 생성자에서 피연산자의 숫자 유무를 판단하여 예외처리하는 기능을 추가하면서, 문제가 되었던 `validateOperand()` 메서드는 삭제되었습니다. 😃 

- https://github.com/next-step/kotlin-racingcar/commit/6f50378cf819efd9a7ab773947d239af744b0b05

코멘트 남겨 주신대로 추후 과제를 진행할때는 메서드 네이밍과 접근 제한자를 신경써서 선언하도록 하겠습니다 😊

sah3122 commented on src/main/kotlin/step2/domain/Calculator.kt:
```diff
@@ -0,0 +1,19 @@
+package step2.domain
+
+enum class Calculator(val operator: String, val operation: (Operand, Operand) -> Int) {
```
깔끔한 enum 정의 좋습니다 😄

namonak replied:
칭찬 감사합니다 😃

sah3122 commented on src/main/kotlin/step2/domain/Operand.kt:
```diff
@@ -0,0 +1,19 @@
+    val value: Int = operand
+
+    init {
+        require(operand >= 0) { "음수는 입력할 수 없습니다." }
```
이번 미션 요구사항이 아니였는데 제가 착각했습니다. 😢

namonak replied:
아닙니다! 확인해 주셔서 감사합니다! 😄

sah3122 commented on src/main/kotlin/step2/domain/Operand.kt:
```diff
@@ -0,0 +1,19 @@
+package step2.domain
+
+class Operand(operand: Int) {
```
피연산자 클래스 정의 👍 
다양한 생성자를 정의한다면 String to Int 를 파싱하는 책임도 Operand 에서 맡을수 있을것 같네요.

https://kotlinlang.org/docs/classes.html#constructors

namonak replied:
전달해 주신 링크 참조하여 문자열 형태로 입력된 피연산자를 예외처리 후 파싱하는 보조 생성자를 추가하였습니다! 😃 

- https://github.com/next-step/kotlin-racingcar/commit/6f50378cf819efd9a7ab773947d239af744b0b05

sah3122 commented on src/main/kotlin/step2/domain/Operand.kt:
```diff
@@ -0,0 +1,19 @@
+        require(operand >= 0) { "음수는 입력할 수 없습니다." }
+    }
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (javaClass != other?.javaClass) return false
+        return true
+    }
+
+    override fun hashCode(): Int {
+        return javaClass.hashCode()
+    }
```
equals & hashCode 를 ovveride 하기 보다는 data class 를 활용해보는건 어떠세요 ? 
https://kotlinlang.org/docs/data-classes.html

namonak replied:
전달해 주신 링크 참고하여 data 클래스로 변경하니, 컴파일러에서 `hashCode()`, `copy()`, `equals()`, `toString()`, `componentsN()` 메서드들을 자동으로 파생해 주어 코드가 훨씬 간결해 졌습니다. 😃 

- https://github.com/next-step/kotlin-racingcar/commit/964ea1cb0eb31164dfe451febf11846f02a81cc5

sah3122 reviewed (approved):
피드백 반영 잘해주셨습니다 👍 
몇가지 고민거리를 남겨두었는데 다음 미션을 진행하면서 반영 부탁드려요 🙏
